### ISSUE TITLE
Create a "server side sync hooks" system

### ISSUE THREAD

**Issue by phenixdotnet** [argoproj/argo-cd#12475]
# Summary

Hi.
First thanks to all contributors for this amazing project ! This a lot of work and a really good project :+1:

I wonder if "server side sync hooks" has been already discuted or not ? There is some cases where it could be very useful to be able to do some tasks before the sync is really started and / or after the sync.

# Motivation

One use case I have is to be able to manipulate infrastructure and some internals systems at the same time the sync is started. For example, when I start to sync a new version of an app, I would like to be able to take a DB snapshot, put the app in maintenance mode and downscale the replicas to 0 before the helm PreSync hook (which contains the DB migration job) start. Those kind of task couldn't be done (or at least I don't want it to be done :) ) on the target cluster side as it require priviliged credentials and accesses.

Another use case is when the app start to sync, I would like to be able to update an internal system which is on a restricted VPC (I use a star VPC architecture with argocd deployed at the center) so the target cluster couldn't be used to contact this system.

# Proposal

Implement 2 "server side sync hooks", one called just before the sync start and one called just after sync finish.

Those hooks could be a grpc call to an external server and 

should

- contains the app manifest

- contains the target cluster credentials and access information

could:
- mutuate the app manifest (may be)
- fail the sync

Not sure if many hooks of the same type should be supported or if only one pre and one post sync hook will be ok

Some questions:
- What about security ? Should those hooks authenticated ?
- What about performance ? Those hooks should timeout, this timeout should be configurable
- What about long running hooks ? Are they authorized (for example a DB snapshoot could take ~5 minutes) ? How the user will know what is in progress ?

If this proposal got some traction, I can try to work on it

Thanks !

**Comment by todaywasawesome** [comment]
@phenixdotnet this is already in Argo CD https://argo-cd.readthedocs.io/en/stable/user-guide/resource_hooks/

Please reopen if I didn't understand.

**Comment by phenixdotnet** [comment]
Hi @todaywasawesome 

I'm totally aware about the resource hooks, sync phase and wave, all those items are great but are executed on the target cluster and not on the argo server cluster. As I say in my first post, I'm looking for a way to "execute something" on the argocd cluster

The best way I can describe what I'm thinking about is the mutating webhook and validation webhook in k8s. Both are privileged component which can react to a resource creation and / or change it.

I agree the "server side sync hook" could be confusing with "resource hooks", if you have a better way to describe it I will be very happy to use it

**Comment by phenixdotnet** [comment]
Hi @todaywasawesome 
Could you please re-open this issue as it is not the same thing than resource hooks nor sync phase and wave ?

Thanks :)

**Comment by todaywasawesome** [comment]
@phenixdotnet Ok, I will reopen it. However, I think there is a great way to achieve this already. 

1) [Enable app health checks in Argo CD](https://argo-cd.readthedocs.io/en/stable/operator-manual/health/#argocd-app)
1) Create a parent application that contains apps you want to sync to the Argo CD instance and apps that you want to sync to external clusters.
1) Set the order in which you want these applications to sync, this become blocking operations so if a sync is not completed from App 1, the sync for App 2 cannot start. 

Here's a [great guide on app dependency management](https://codefresh.io/blog/argo-cd-application-dependencies/) from @christianh814.

Adding support for sync hooks to select different clusters than the destination cluster feels like adding a lot of complexity for not much benefit. I might be totally alone there and lacking imagination so feel free to keep pushing ahead! 😃

**Comment by phenixdotnet** [comment]
Thanks you for reopening the issue.


I think there is a misunderstanding

I'm not trying to sync different apps on different clusters, I'm trying to handle the sync event at the argocd server side because do that allow the operator to validate, mutate and react to this event in a "privileged" environment (And this notion of privileged environment is really important because for security reason, I really don't want to have "admin" credentials sent to a "standard" cluster nor having IAM roles and service accounts allowed to manipulate the infrastructure or k8s objects).

From my understanding, using the apps of apps pattern doesn't guarantee the 2 apps will be synced "on the same shoot". 
If, for example, the first app take a DB snapshot (with the idea to be able to use it to restore the db if the migrations aren't running well) and the dependent app is synced "only" 1 hour after, this could be ... well "complicated".

**Comment by christianh814** [comment]
> Thanks you for reopening the issue.
> 
> I think there is a misunderstanding
> 
> I'm not trying to sync different apps on different clusters, I'm trying to handle the sync event at the argocd server side because do that allow the operator to validate, mutate and react to this event in a "privileged" environment (And this notion of privileged environment is really important because for security reason, I really don't want to have "admin" credentials sent to a "standard" cluster nor having IAM roles and service accounts allowed to manipulate the infrastructure or k8s objects).

Are you saying you want to validate YAML/Resouces on the Argo CD instance itself rather than the host cluster?

If so that's not possible and that's not an "Argo CD" issue either. The CRDs may exist on the managed cluster but not on the cluster where Argo CD is running. Which is more of a "Kubernetes issue" than an Argo CD one.

Also, you can also validate things via plugins.

> 
> From my understanding, using the apps of apps pattern doesn't guarantee the 2 apps will be synced "on the same shoot". If, for example, the first app take a DB snapshot (with the idea to be able to use it to restore the db if the migrations aren't running well) and the dependent app is synced "only" 1 hour after, this could be ... well "complicated".

This can be achived with ApplicationSets Progressive Syncs + Using something like Argo Rollouts. It's "complicated" because Argo CD doesn't automagically do this...you'll need to integrate other tools.


Please clarify if I got the question/requestion wrong.

**Comment by phenixdotnet** [comment]
Hi @christianh814 

> Are you saying you want to validate YAML/Resources on the Argo CD instance itself rather than the host cluster?

Not the resources, the actual validation and diff mechanisms are fine for me, but I would like to be able to validate the context around the deployment. Deploying a new version of an app could be more than "just" apply yaml and resources, you can have a lot of different constrains, including QA and/or regulatory validation or infrastructure check

The plugins are a great mechanism to customize resources generation but they are not involved at the sync time. So between the moment where the plugin is called and the moment where the resources are applied, the context could change and allow or not the sync.

And the new extension mechanism looks great but it seem not be able to replace the sync action used by the UI and CLI. And this is an issue IMHO because this lead to an additional complexity for the user.

The best way I can describe what I'm looking to do is this little schema
![Blank diagram (1)](https://user-images.githubusercontent.com/2082554/219870943-b0d31720-f3cb-4e67-b186-914d09a1dd80.png)

In this scenario, the "context validation thing" couldn't be run on the production vpc because it will not have access to the required servers.


I don't want boring you with this idea, if you're still not convinced it's a good feature candidate, I will close this issue :)

Regards

**Comment by todaywasawesome** [comment]
For something like a quality gate I've used OPA Gatekeeper and Kyverno to check attestations. This works especially well with Argo CD >2.4 because of server side apply. 

I think you should leave the issue open a bit to see if other folks jump on, or bring it up in the contributor office hours.

**Comment by phenixdotnet** [comment]
Yes OPA Gatekeeper and kyverno are great tools but they run on the target cluster which is unfortunately not suitable with my needs.

> I think you should leave the issue open a bit to see if other folks jump on, or bring it up in the contributor office hours.
okay :+1:

**Comment by crenshaw-dev** [comment]
I like this idea. A simplified expression is "I would like to do things on the Argo CD cluster via the Application API." Hooks are just one example of a way you could use the Application API to interface with the Argo CD cluster. You could also gitops-deploy things to the host cluster, have resources from the host cluster appear in the resource tree, and run actions against resources in the host cluster.

This would be a _big_ change and would have massive implications for security in Argo CD. But I think it would unlock a variety of really interesting use cases.

**Comment by Voigtus** [comment]
Just my 2 cents, why I think this is a splendid idea:
It would enable the possibility to have centralised "pre-flight" checks for syncs. For example to verify that all regulatory constraints are fulfilled, before a deployment can happen, and "pull the plug" if necessary, while at the same time no one can weasel their way around it.

Or create ressources out of the argoCD context, with privileged users, that are not available to users of ArgoCD depending on the ressources that are changed in the upcoming sync.

Or (yes I know, quite a anti-pattern, but I think there might be people who would want to do that) enable centralised "sync windows", with a remote server that tells the application "yes, you are allowed to be synced", but is controlled by an external instance and not by the team, that created the ArgoCD application.
